\documentclass[11pt]{article}

    \usepackage[breakable]{tcolorbox}
    \usepackage{parskip} % Stop auto-indenting (to mimic markdown behaviour)
    
    \usepackage{iftex}
    \ifPDFTeX
    	\usepackage[T1]{fontenc}
    	\usepackage{mathpazo}
    \else
    	\usepackage{fontspec}
    \fi

    % Basic figure setup, for now with no caption control since it's done
    % automatically by Pandoc (which extracts ![](path) syntax from Markdown).
    \usepackage{graphicx}
    % Maintain compatibility with old templates. Remove in nbconvert 6.0
    \let\Oldincludegraphics\includegraphics
    % Ensure that by default, figures have no caption (until we provide a
    % proper Figure object with a Caption API and a way to capture that
    % in the conversion process - todo).
    \usepackage{caption}
    \DeclareCaptionFormat{nocaption}{}
    \captionsetup{format=nocaption,aboveskip=0pt,belowskip=0pt}

    \usepackage[Export]{adjustbox} % Used to constrain images to a maximum size
    \adjustboxset{max size={0.9\linewidth}{0.9\paperheight}}
    \usepackage{float}
    \floatplacement{figure}{H} % forces figures to be placed at the correct location
    \usepackage{xcolor} % Allow colors to be defined
    \usepackage{enumerate} % Needed for markdown enumerations to work
    \usepackage{geometry} % Used to adjust the document margins
    \usepackage{amsmath} % Equations
    \usepackage{amssymb} % Equations
    \usepackage{textcomp} % defines textquotesingle
    % Hack from http://tex.stackexchange.com/a/47451/13684:
    \AtBeginDocument{%
        \def\PYZsq{\textquotesingle}% Upright quotes in Pygmentized code
    }
    \usepackage{upquote} % Upright quotes for verbatim code
    \usepackage{eurosym} % defines \euro
    \usepackage[mathletters]{ucs} % Extended unicode (utf-8) support
    \usepackage{fancyvrb} % verbatim replacement that allows latex
    \usepackage{grffile} % extends the file name processing of package graphics 
                         % to support a larger range
    \makeatletter % fix for grffile with XeLaTeX
    \def\Gread@@xetex#1{%
      \IfFileExists{"\Gin@base".bb}%
      {\Gread@eps{\Gin@base.bb}}%
      {\Gread@@xetex@aux#1}%
    }
    \makeatother

    % The hyperref package gives us a pdf with properly built
    % internal navigation ('pdf bookmarks' for the table of contents,
    % internal cross-reference links, web links for URLs, etc.)
    \usepackage{hyperref}
    % The default LaTeX title has an obnoxious amount of whitespace. By default,
    % titling removes some of it. It also provides customization options.
    \usepackage{titling}
    \usepackage{longtable} % longtable support required by pandoc >1.10
    \usepackage{booktabs}  % table support for pandoc > 1.12.2
    \usepackage[inline]{enumitem} % IRkernel/repr support (it uses the enumerate* environment)
    \usepackage[normalem]{ulem} % ulem is needed to support strikethroughs (\sout)
                                % normalem makes italics be italics, not underlines
    \usepackage{mathrsfs}
    

    
    % Colors for the hyperref package
    \definecolor{urlcolor}{rgb}{0,.145,.698}
    \definecolor{linkcolor}{rgb}{.71,0.21,0.01}
    \definecolor{citecolor}{rgb}{.12,.54,.11}

    % ANSI colors
    \definecolor{ansi-black}{HTML}{3E424D}
    \definecolor{ansi-black-intense}{HTML}{282C36}
    \definecolor{ansi-red}{HTML}{E75C58}
    \definecolor{ansi-red-intense}{HTML}{B22B31}
    \definecolor{ansi-green}{HTML}{00A250}
    \definecolor{ansi-green-intense}{HTML}{007427}
    \definecolor{ansi-yellow}{HTML}{DDB62B}
    \definecolor{ansi-yellow-intense}{HTML}{B27D12}
    \definecolor{ansi-blue}{HTML}{208FFB}
    \definecolor{ansi-blue-intense}{HTML}{0065CA}
    \definecolor{ansi-magenta}{HTML}{D160C4}
    \definecolor{ansi-magenta-intense}{HTML}{A03196}
    \definecolor{ansi-cyan}{HTML}{60C6C8}
    \definecolor{ansi-cyan-intense}{HTML}{258F8F}
    \definecolor{ansi-white}{HTML}{C5C1B4}
    \definecolor{ansi-white-intense}{HTML}{A1A6B2}
    \definecolor{ansi-default-inverse-fg}{HTML}{FFFFFF}
    \definecolor{ansi-default-inverse-bg}{HTML}{000000}

    % commands and environments needed by pandoc snippets
    % extracted from the output of `pandoc -s`
    \providecommand{\tightlist}{%
      \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
    \DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
    % Add ',fontsize=\small' for more characters per line
    \newenvironment{Shaded}{}{}
    \newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{\textbf{{#1}}}}
    \newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.56,0.13,0.00}{{#1}}}
    \newcommand{\DecValTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\FloatTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\CharTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\StringTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\CommentTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textit{{#1}}}}
    \newcommand{\OtherTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{{#1}}}
    \newcommand{\AlertTok}[1]{\textcolor[rgb]{1.00,0.00,0.00}{\textbf{{#1}}}}
    \newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.02,0.16,0.49}{{#1}}}
    \newcommand{\RegionMarkerTok}[1]{{#1}}
    \newcommand{\ErrorTok}[1]{\textcolor[rgb]{1.00,0.00,0.00}{\textbf{{#1}}}}
    \newcommand{\NormalTok}[1]{{#1}}
    
    % Additional commands for more recent versions of Pandoc
    \newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.53,0.00,0.00}{{#1}}}
    \newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.73,0.40,0.53}{{#1}}}
    \newcommand{\ImportTok}[1]{{#1}}
    \newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.73,0.13,0.13}{\textit{{#1}}}}
    \newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\VariableTok}[1]{\textcolor[rgb]{0.10,0.09,0.49}{{#1}}}
    \newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{\textbf{{#1}}}}
    \newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.40,0.40,0.40}{{#1}}}
    \newcommand{\BuiltInTok}[1]{{#1}}
    \newcommand{\ExtensionTok}[1]{{#1}}
    \newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.74,0.48,0.00}{{#1}}}
    \newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.49,0.56,0.16}{{#1}}}
    \newcommand{\InformationTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\WarningTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    
    
    % Define a nice break command that doesn't care if a line doesn't already
    % exist.
    \def\br{\hspace*{\fill} \\* }
    % Math Jax compatibility definitions
    \def\gt{>}
    \def\lt{<}
    \let\Oldtex\TeX
    \let\Oldlatex\LaTeX
    \renewcommand{\TeX}{\textrm{\Oldtex}}
    \renewcommand{\LaTeX}{\textrm{\Oldlatex}}
    % Document parameters
    % Document title
    \title{fastp}
    
    
    
    
    
% Pygments definitions
\makeatletter
\def\PY@reset{\let\PY@it=\relax \let\PY@bf=\relax%
    \let\PY@ul=\relax \let\PY@tc=\relax%
    \let\PY@bc=\relax \let\PY@ff=\relax}
\def\PY@tok#1{\csname PY@tok@#1\endcsname}
\def\PY@toks#1+{\ifx\relax#1\empty\else%
    \PY@tok{#1}\expandafter\PY@toks\fi}
\def\PY@do#1{\PY@bc{\PY@tc{\PY@ul{%
    \PY@it{\PY@bf{\PY@ff{#1}}}}}}}
\def\PY#1#2{\PY@reset\PY@toks#1+\relax+\PY@do{#2}}

\expandafter\def\csname PY@tok@w\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.73,0.73}{##1}}}
\expandafter\def\csname PY@tok@c\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cp\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.74,0.48,0.00}{##1}}}
\expandafter\def\csname PY@tok@k\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kp\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kt\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.69,0.00,0.25}{##1}}}
\expandafter\def\csname PY@tok@o\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@ow\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.67,0.13,1.00}{##1}}}
\expandafter\def\csname PY@tok@nb\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@nf\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@nc\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@nn\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@ne\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.82,0.25,0.23}{##1}}}
\expandafter\def\csname PY@tok@nv\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@no\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.53,0.00,0.00}{##1}}}
\expandafter\def\csname PY@tok@nl\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.63,0.63,0.00}{##1}}}
\expandafter\def\csname PY@tok@ni\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.60,0.60,0.60}{##1}}}
\expandafter\def\csname PY@tok@na\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.49,0.56,0.16}{##1}}}
\expandafter\def\csname PY@tok@nt\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@nd\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.67,0.13,1.00}{##1}}}
\expandafter\def\csname PY@tok@s\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sd\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@si\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.53}{##1}}}
\expandafter\def\csname PY@tok@se\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.13}{##1}}}
\expandafter\def\csname PY@tok@sr\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.53}{##1}}}
\expandafter\def\csname PY@tok@ss\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@sx\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@m\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@gh\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\expandafter\def\csname PY@tok@gu\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.50,0.00,0.50}{##1}}}
\expandafter\def\csname PY@tok@gd\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.63,0.00,0.00}{##1}}}
\expandafter\def\csname PY@tok@gi\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.63,0.00}{##1}}}
\expandafter\def\csname PY@tok@gr\endcsname{\def\PY@tc##1{\textcolor[rgb]{1.00,0.00,0.00}{##1}}}
\expandafter\def\csname PY@tok@ge\endcsname{\let\PY@it=\textit}
\expandafter\def\csname PY@tok@gs\endcsname{\let\PY@bf=\textbf}
\expandafter\def\csname PY@tok@gp\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\expandafter\def\csname PY@tok@go\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.53,0.53,0.53}{##1}}}
\expandafter\def\csname PY@tok@gt\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.27,0.87}{##1}}}
\expandafter\def\csname PY@tok@err\endcsname{\def\PY@bc##1{\setlength{\fboxsep}{0pt}\fcolorbox[rgb]{1.00,0.00,0.00}{1,1,1}{\strut ##1}}}
\expandafter\def\csname PY@tok@kc\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kd\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kn\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kr\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@bp\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@fm\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@vc\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@vg\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@vi\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@vm\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@sa\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sb\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sc\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@dl\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@s2\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sh\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@s1\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@mb\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mf\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mh\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mi\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@il\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mo\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@ch\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cm\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cpf\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@c1\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cs\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}

\def\PYZbs{\char`\\}
\def\PYZus{\char`\_}
\def\PYZob{\char`\{}
\def\PYZcb{\char`\}}
\def\PYZca{\char`\^}
\def\PYZam{\char`\&}
\def\PYZlt{\char`\<}
\def\PYZgt{\char`\>}
\def\PYZsh{\char`\#}
\def\PYZpc{\char`\%}
\def\PYZdl{\char`\$}
\def\PYZhy{\char`\-}
\def\PYZsq{\char`\'}
\def\PYZdq{\char`\"}
\def\PYZti{\char`\~}
% for compatibility with earlier versions
\def\PYZat{@}
\def\PYZlb{[}
\def\PYZrb{]}
\makeatother


    % For linebreaks inside Verbatim environment from package fancyvrb. 
    \makeatletter
        \newbox\Wrappedcontinuationbox 
        \newbox\Wrappedvisiblespacebox 
        \newcommand*\Wrappedvisiblespace {\textcolor{red}{\textvisiblespace}} 
        \newcommand*\Wrappedcontinuationsymbol {\textcolor{red}{\llap{\tiny$\m@th\hookrightarrow$}}} 
        \newcommand*\Wrappedcontinuationindent {3ex } 
        \newcommand*\Wrappedafterbreak {\kern\Wrappedcontinuationindent\copy\Wrappedcontinuationbox} 
        % Take advantage of the already applied Pygments mark-up to insert 
        % potential linebreaks for TeX processing. 
        %        {, <, #, %, $, ' and ": go to next line. 
        %        _, }, ^, &, >, - and ~: stay at end of broken line. 
        % Use of \textquotesingle for straight quote. 
        \newcommand*\Wrappedbreaksatspecials {% 
            \def\PYGZus{\discretionary{\char`\_}{\Wrappedafterbreak}{\char`\_}}% 
            \def\PYGZob{\discretionary{}{\Wrappedafterbreak\char`\{}{\char`\{}}% 
            \def\PYGZcb{\discretionary{\char`\}}{\Wrappedafterbreak}{\char`\}}}% 
            \def\PYGZca{\discretionary{\char`\^}{\Wrappedafterbreak}{\char`\^}}% 
            \def\PYGZam{\discretionary{\char`\&}{\Wrappedafterbreak}{\char`\&}}% 
            \def\PYGZlt{\discretionary{}{\Wrappedafterbreak\char`\<}{\char`\<}}% 
            \def\PYGZgt{\discretionary{\char`\>}{\Wrappedafterbreak}{\char`\>}}% 
            \def\PYGZsh{\discretionary{}{\Wrappedafterbreak\char`\#}{\char`\#}}% 
            \def\PYGZpc{\discretionary{}{\Wrappedafterbreak\char`\%}{\char`\%}}% 
            \def\PYGZdl{\discretionary{}{\Wrappedafterbreak\char`\$}{\char`\$}}% 
            \def\PYGZhy{\discretionary{\char`\-}{\Wrappedafterbreak}{\char`\-}}% 
            \def\PYGZsq{\discretionary{}{\Wrappedafterbreak\textquotesingle}{\textquotesingle}}% 
            \def\PYGZdq{\discretionary{}{\Wrappedafterbreak\char`\"}{\char`\"}}% 
            \def\PYGZti{\discretionary{\char`\~}{\Wrappedafterbreak}{\char`\~}}% 
        } 
        % Some characters . , ; ? ! / are not pygmentized. 
        % This macro makes them "active" and they will insert potential linebreaks 
        \newcommand*\Wrappedbreaksatpunct {% 
            \lccode`\~`\.\lowercase{\def~}{\discretionary{\hbox{\char`\.}}{\Wrappedafterbreak}{\hbox{\char`\.}}}% 
            \lccode`\~`\,\lowercase{\def~}{\discretionary{\hbox{\char`\,}}{\Wrappedafterbreak}{\hbox{\char`\,}}}% 
            \lccode`\~`\;\lowercase{\def~}{\discretionary{\hbox{\char`\;}}{\Wrappedafterbreak}{\hbox{\char`\;}}}% 
            \lccode`\~`\:\lowercase{\def~}{\discretionary{\hbox{\char`\:}}{\Wrappedafterbreak}{\hbox{\char`\:}}}% 
            \lccode`\~`\?\lowercase{\def~}{\discretionary{\hbox{\char`\?}}{\Wrappedafterbreak}{\hbox{\char`\?}}}% 
            \lccode`\~`\!\lowercase{\def~}{\discretionary{\hbox{\char`\!}}{\Wrappedafterbreak}{\hbox{\char`\!}}}% 
            \lccode`\~`\/\lowercase{\def~}{\discretionary{\hbox{\char`\/}}{\Wrappedafterbreak}{\hbox{\char`\/}}}% 
            \catcode`\.\active
            \catcode`\,\active 
            \catcode`\;\active
            \catcode`\:\active
            \catcode`\?\active
            \catcode`\!\active
            \catcode`\/\active 
            \lccode`\~`\~ 	
        }
    \makeatother

    \let\OriginalVerbatim=\Verbatim
    \makeatletter
    \renewcommand{\Verbatim}[1][1]{%
        %\parskip\z@skip
        \sbox\Wrappedcontinuationbox {\Wrappedcontinuationsymbol}%
        \sbox\Wrappedvisiblespacebox {\FV@SetupFont\Wrappedvisiblespace}%
        \def\FancyVerbFormatLine ##1{\hsize\linewidth
            \vtop{\raggedright\hyphenpenalty\z@\exhyphenpenalty\z@
                \doublehyphendemerits\z@\finalhyphendemerits\z@
                \strut ##1\strut}%
        }%
        % If the linebreak is at a space, the latter will be displayed as visible
        % space at end of first line, and a continuation symbol starts next line.
        % Stretch/shrink are however usually zero for typewriter font.
        \def\FV@Space {%
            \nobreak\hskip\z@ plus\fontdimen3\font minus\fontdimen4\font
            \discretionary{\copy\Wrappedvisiblespacebox}{\Wrappedafterbreak}
            {\kern\fontdimen2\font}%
        }%
        
        % Allow breaks at special characters using \PYG... macros.
        \Wrappedbreaksatspecials
        % Breaks at punctuation characters . , ; ? ! and / need catcode=\active 	
        \OriginalVerbatim[#1,codes*=\Wrappedbreaksatpunct]%
    }
    \makeatother

    % Exact colors from NB
    \definecolor{incolor}{HTML}{303F9F}
    \definecolor{outcolor}{HTML}{D84315}
    \definecolor{cellborder}{HTML}{CFCFCF}
    \definecolor{cellbackground}{HTML}{F7F7F7}
    
    % prompt
    \makeatletter
    \newcommand{\boxspacing}{\kern\kvtcb@left@rule\kern\kvtcb@boxsep}
    \makeatother
    \newcommand{\prompt}[4]{
        \ttfamily\llap{{\color{#2}[#3]:\hspace{3pt}#4}}\vspace{-\baselineskip}
    }
    

    
    % Prevent overflowing lines due to hard-to-break entities
    \sloppy 
    % Setup hyperref package
    \hypersetup{
      breaklinks=true,  % so long urls are correctly broken across lines
      colorlinks=true,
      urlcolor=urlcolor,
      linkcolor=linkcolor,
      citecolor=citecolor,
      }
    % Slightly bigger margins than the latex defaults
    
    \geometry{verbose,tmargin=1in,bmargin=1in,lmargin=1in,rmargin=1in}
    
    

\begin{document}
    
    \maketitle
    
    

    
    \hypertarget{raw-read-structure}{%
\subsection{Raw read structure}\label{raw-read-structure}}

A paired-end read is a PCR-amplified fragment enclosed inside adapters.
The whole fragment is sequenced.

\hypertarget{raw-sequence-quality}{%
\subsection{Raw Sequence Quality}\label{raw-sequence-quality}}

\textbf{Phred quality score} (Q) was originally developed by the program
Phred to measure base-calling reliability from Sanger sequencing
chromatograms.

\(Q = -10 * log_{10}(P)\) where P is the probability of erroneous base
calling.

\textbf{For example:} * Q20: \(-10 * log_{10}(1/100) = 20\) * Q30:
\(-10 * log_{10}(1/1000) = 30\) * Q40: \(-10 * log_{10}(1/10^4) = 40\) *
Q50: \(-10 * log_{10}(10^{-5}) = 50\) * Q60:
\(-10 * log_{10}(10^{-6}) = 60\)

\textbf{Quality scoring range:}

\begin{itemize}
\tightlist
\item
  \(< 20\): poor quality
\item
  \(20 - 30\): moderate quality
\item
  \(> 30\): high quality
\end{itemize}

\textbf{Example fastq read}

\texttt{@SRR5832182.1\ HISEQ-MFG:180:hw5f3bcxx:2:1101:1205:2195\ length=101\ CGCGAACAGTCCTCAAATCGGGGATCGAAAAAGGATTCGTGGAATCACAGTATTTTTATGTATAATTTCTTCCACCGAAGAGGGAGCAACAAGAGAGCGGC\ +SRR5832182.1\ HISEQ-MFG:180:hw5f3bcxx:2:1101:1205:2195\ length=101\ DDDDDHHIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIGHHHIEHIIIIIIIIIIFHIIIHIIIIIHIIIIIIIIIIIIIIIIGHHIIIIIHIIHIIIIH}

Each read (in fastq format) has four lines: 1. ID 2. Base call (the
sequence) 3. ID again for some reason (or just a plus sign) 4. Quality
symbol for each base call

\hypertarget{quality-symbols-for-q-32-40}{%
\subsection{Quality symbols for Q:
32-40}\label{quality-symbols-for-q-32-40}}

\begin{longtable}[]{@{}lll@{}}
\toprule
symbol & ascii value & Q\tabularnewline
\midrule
\endhead
A & 65 & 32\tabularnewline
B & 66 & 33\tabularnewline
C & 67 & 34\tabularnewline
D & 68 & 35\tabularnewline
E & 69 & 36\tabularnewline
F & 70 & 37\tabularnewline
G & 71 & 38\tabularnewline
H & 72 & 39\tabularnewline
I & 73 & 40\tabularnewline
\bottomrule
\end{longtable}

Link for full(er) table:
https://support.illumina.com/help/BaseSpace\_OLH\_009008/Content/Source/Informatics/BS/QualityScoreEncoding\_swBS.htm

Is \(ascii -33\), going all the way to 126 (symbol=`\textasciitilde{}',
Q=93). See http://www.asciitable.com/ if you're interested in how this
works.

\hypertarget{quality-scores-for-a-sequencing-run}{%
\subsection{Quality scores for a sequencing
run}\label{quality-scores-for-a-sequencing-run}}

\textbf{Li et al.~Fig 1. a)} Parallel boxplot showing ``per nucleotide
quality score.'' All reads are overlaid together, and then summarize
Phred quality score (Y-axis) for each position of read from 5′ to 3′ end
(X-axis). \textbf{(b)} ``per sequence quality score'' distribution. For
each read, ``per sequence quality score'' is calculated as the average
Phred quality score (X-axis) across all nucleotides

\hypertarget{nucleotide-composition-and-gc-content}{%
\subsection{Nucleotide Composition and GC
Content}\label{nucleotide-composition-and-gc-content}}

\textbf{Li et al.~Fig 2. (a)} Diagram showing nucleotide composition
bias at the beginning of reads. All reads are overlaid together, and
then calculate nucleotide frequency (Y-axis) for each position of read
(X-axis). Four nucleotides were indicated using different colors.
\textbf{(b)} ``per sequence GC content'' distribution. This normal
distribution should be centered at the GC content of your coding genome
(exons only). A shift represents a bias or contamination.

\hypertarget{fastp}{%
\section{fastp}\label{fastp}}

\textbf{Reference}: https://watermark.silverchair.com/bty560.pdf fastp:
an ultra-fast all-in-one FASTQ preprocessor \textbf{Bioinformatics,} 34,
2018, i884--i890

\textbf{Today's reading}:
http://rna.colostate.edu/2019/lib/exe/fetch.php?media=assignments:li\_etal\_qualitycontrol\_2015.pdf

\hypertarget{benefits-of-fastp-over-other-tools}{%
\subsection{Benefits of fastp over other
tools}\label{benefits-of-fastp-over-other-tools}}

\begin{itemize}
\tightlist
\item
  Trimming and Quality in one
\item
  Faster implementation
\item
  Automatically finds adapters
\item
  Interactive HTML output
\end{itemize}

\hypertarget{fastp-reports}{%
\section{fastp reports}\label{fastp-reports}}

\hypertarget{adapter-identification}{%
\subsection{Adapter identification}\label{adapter-identification}}

\ldots{}

\textbf{\emph{Courtesy of Rob Williams}}

\hypertarget{pcr-duplication}{%
\subsection{PCR duplication}\label{pcr-duplication}}

\textbf{\emph{Courtesy of Rob Williams}}

\hypertarget{base-content-filtering}{%
\subsection{Base content filtering}\label{base-content-filtering}}

\textbf{\emph{fastp paper figure 3}}

\begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

    \hypertarget{lets-process-some-data}{%
\section{Let's process some data}\label{lets-process-some-data}}

\hypertarget{where-are-we-right-now}{%
\subsection{Where are we right now?}\label{where-are-we-right-now}}

\begin{itemize}
\tightlist
\item
  We're already on summit
\item
  We're already running a job (a 12 hour interactive job that we
  \emph{spawned}).
\item
  That means we're on a compute node, and it's OK to run a full job
  (we're usually on a login or compile node).
\item
  Counts against fairshare. So we need to stop the server when we're
  done. (See the bottom of the notebook and/or the wiki page).
\item
  We're in the jupyterhub web-based interface instead of the basic
  terminal interface.

  \begin{itemize}
  \tightlist
  \item
    File menu, Edit menu, etc. not part of your OS, but inside your web
    browser.
  \item
    Open tabs are on the right - these are notebooks, consoles, and
    other (including terminals).
  \item
    Icons below files menu:

    \begin{itemize}
    \tightlist
    \item
      Plus sign - Launcher
    \item
      Dark Folder with light plus sign - new Folder
    \item
      Up-arrow with underline - upload file
    \item
      Circular arrow - refresh file list
    \end{itemize}
  \item
    next row of icons: current location:
    \texttt{{[}folder\ icon{]}/.../parent\ directory/current\ directory}
  \item
    file browser - \textbf{Name} \textbf{Last Modified}
  \end{itemize}
\end{itemize}

\hypertarget{workflow}{%
\subsection{Workflow}\label{workflow}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Create a working directory for this exercise.
\item
  Link to data (I already downloaded it - very large).
\item
  Set up \emph{fastp} through a container and make an alias.
\item
  Run the program on a single dataset. Inspect.
\item
  Write a script to run the other datasets in a batch job (we'll do this
  from the terminal)
\end{enumerate}

First, let's set aside some space in the working directory.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{c+c1}{\PYZsh{} Step 1: create a workspace}
\PY{n+nb}{cd} /scratch/summit/\PY{n+nv}{\PYZdl{}USER}
mkdir DSCI512\PYZus{}RNAseq\PYZus{}19\PYZus{}19\PYZus{}11
\PY{n+nb}{cd} DSCI512\PYZus{}RNAseq\PYZus{}19\PYZus{}19\PYZus{}11
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{n+nb}{pwd}
\end{Verbatim}
\end{tcolorbox}

    In the upper lefthand corner, go to File -\textgreater{}
\texttt{Open\ from\ path...} and paste in the output from the previous
command.

This will open the file browser to the current location. *** Now we're
going to link a data directory to this current directory. We do this
because the data is very large and will take too long for a
demonstration. The link will reside in the present directory and act
like any other, except you won't be able to change its contents. It is
\textbf{read-only.}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{c+c1}{\PYZsh{} make directories to use through processing}
\PY{c+c1}{\PYZsh{} skipping 01\PYZus{}input \PYZhy{} we will make that with a link below}
mkdir 02\PYZus{}output
mkdir 03\PYZus{}scripts
mkdir 04\PYZus{}logs
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{c+c1}{\PYZsh{}link to the data directory (I have already downloaded everything)}
ln \PYZhy{}sv /scratch/summit/dcking@colostate.edu/DSCI512/2019/data 01\PYZus{}input
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{c+c1}{\PYZsh{} Look at your directory structure.}
ls \PYZhy{}lh
\end{Verbatim}
\end{tcolorbox}

    🔎 Navigate around the file browser on the left * doubleclick a link to
descend * go to a parent directory by clicking the name after the 3 dots
in the location field:
\texttt{folder\_icon/.../parent\_directory/current\_directory}

\hypertarget{breakdown-of-data-in-01_input}{%
\paragraph{Breakdown of data in
01\_input/}\label{breakdown-of-data-in-01_input}}

\begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

\textbf{Experimental design}

\begin{longtable}[]{@{}lll@{}}
\toprule
IDs (3 reps each) & Temperature & Food Source\tabularnewline
\midrule
\endhead
SRR5832182-SRR5832184 & 15°C & \emph{E. coli}\tabularnewline
SRR5832185-SRR5832187 & 20°C & \emph{E. coli}\tabularnewline
SRR5832188-SRR5832190 & 25°C & \emph{E. coli}\tabularnewline
SRR5832191-SRR5832193 & 15°C & \emph{B. subtilis}\tabularnewline
SRR5832194-SRR5832196 & 20°C & \emph{B. subtilis}\tabularnewline
SRR5832197-SRR5832199 & 25°C & \emph{B. subtilis}\tabularnewline
\bottomrule
\end{longtable}

\begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

\begin{itemize}
\tightlist
\item
  6 conditions
\item
  3 reps each
\item
  Each replicate is paired-end. Thus, \emph{SRR5832182} is represented
  by:

  \begin{itemize}
  \tightlist
  \item
    SRR5832182\_1.fastq - ``left'' end reads.
  \item
    SRR5832182\_2.fastq - ``right'' end reads.
  \end{itemize}
\item
  36 total files
\end{itemize}

\hypertarget{running-fastp}{%
\section{Running fastp}\label{running-fastp}}

We will run this through a singularity container:

\begin{itemize}
\tightlist
\item
  Load the singularity module
\item
  Test the container with the full command (long)
\item
  Make an alias for the long command
\end{itemize}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{c+c1}{\PYZsh{} Step 3: load the module that works with containers}
module load singularity
module list
\end{Verbatim}
\end{tcolorbox}

    The following command:
\texttt{singularity\ exec\ /projects/dcking@colostate.edu/containers/Summit\_RNAseq\_container.sif\ fastp}
* \textbf{singularity} - A program that reads a container. *
\textbf{exec} - verb: execute * \textbf{\emph{{[}path to container
image{]}}}: The container itself, called an image. * \textbf{fastp}: The
program you want to execute.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{c+c1}{\PYZsh{} Step 4: Run fastp through the container without arguments\PYZhy{} gives catalog of available flags}
singularity \PY{n+nb}{exec} /projects/dcking@colostate.edu/containers/Summit\PYZus{}RNAseq\PYZus{}container.sif fastp
\end{Verbatim}
\end{tcolorbox}

    \textbf{\emph{This is OK ---\textgreater{}}} WARNING: Non existent `bind
path' source: `/rc\_scratch' \textless--- \textbf{\emph{It's just part
of the configuration.}}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{c+c1}{\PYZsh{} Make a shortcut for fastp}
\PY{n+nv}{fastp}\PY{o}{=}\PY{l+s+s1}{\PYZsq{}singularity exec /projects/dcking@colostate.edu/containers/Summit\PYZus{}RNAseq\PYZus{}container.sif fastp\PYZsq{}}
\end{Verbatim}
\end{tcolorbox}

    You will now be able to type \emph{fastp} in place of the long command.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{c+c1}{\PYZsh{} Test the alias\PYZhy{} same output.}
\PY{n+nv}{\PYZdl{}fastp}
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
wc \PYZhy{}l 01\PYZus{}input/SRR5832182\PYZus{}1.fastq 01\PYZus{}input/SRR5832182\PYZus{}2.fastq
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{fastp-command-usage}{%
\subsection{fastp command usage}\label{fastp-command-usage}}

\textbf{The usage message tells us for \emph{paired end} data:}

\texttt{-i\ readfile\_1.fastq\ -I\ readfile\_2.fastq}

\texttt{-o\ outputfile\_1.fastq\ -O\ outputfile\_2.fastq}

\texttt{{[}options{]}}

For the options:

\begin{itemize}
\tightlist
\item
  \textbf{-x}: remove polyX (polyAs polyCs polyGs polyTs)
\item
  \textbf{-p}: overrepresentation analysis
\item
  \textbf{--thread}: We only have 1 on jupyterhub. We can use more in
  our script.
\item
  \textbf{-h,-j}: The report filenames in html, json (javascript object
  notation).
\end{itemize}

To explicitly specify adapter sequences, see this part of the usage
message:

\begin{verbatim}
-a, --adapter_sequence               the adapter for read1. For SE data, if not specified, the adapter will be auto-detected. For PE data, this is used if R1/R2 are found not overlapped. (string [=auto])
      --adapter_sequence_r2            the adapter for read2 (PE data only). This is used if R1/R2 are found not overlapped. If not specified, it will be the same as <adapter_sequence> (string [=auto])
      --adapter_fasta                  specify a FASTA file to trim both read1 and read2 (if PE) by all the sequences in this FASTA file (string [=])
      --detect_adapter_for_pe          by default, the auto-detection for adapter is for SE data input only, turn on this option to enable it for PE data.
\end{verbatim}

\begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

\hypertarget{command-version-1}{%
\subsubsection{Command version 1:}\label{command-version-1}}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{n+nv}{\PYZdl{}fastp} \PYZhy{}i 01\PYZus{}input/SRR5832199\PYZus{}1.fastq       \PYZhy{}I 01\PYZus{}input/SRR5832199\PYZus{}2.fastq \PY{l+s+se}{\PYZbs{}}
           \PYZhy{}o 02\PYZus{}output/SRR5832199\PYZus{}trim\PYZus{}1.fastq \PYZhy{}O 02\PYZus{}output/SRR5832199\PYZus{}trim\PYZus{}2.fastq \PY{l+s+se}{\PYZbs{}}
           \PYZhy{}h 02\PYZus{}output/SRR5832199\PYZus{}report.html  \PYZhy{}j 02\PYZus{}output/SRR5832199\PYZus{}report.json\PY{l+s+se}{\PYZbs{}}
           \PYZhy{}\PYZhy{}thread \PY{l+m}{1} \PY{l+s+se}{\PYZbs{}}
           \PYZhy{}g \PYZhy{}x \PYZhy{}p
\end{Verbatim}
\end{tcolorbox}

    \textbf{\emph{This is OK ---\textgreater{}}} WARNING: Non existent `bind
path' source: `/rc\_scratch' \textless--- \textbf{\emph{It's just part
of the configuration.}}

⏳ \textbf{Still waiting?} It is running while you still see
\texttt{{[}*{]}} with the asterisk. Give it about a minute.

\hypertarget{notice-the-terminal-output}{%
\subsubsection{Notice the terminal
output}\label{notice-the-terminal-output}}

\begin{itemize}
\tightlist
\item
  Basic statistics - covered in depth by the html output
\item
  Amount of time it ran - this will help us scale up in the batch
  script.
\end{itemize}

\hypertarget{check-the-output}{%
\subsubsection{Check the output}\label{check-the-output}}

Navigate to \textbf{02\_output} in the file browser and double-click
\emph{SRR5832199\_report.html}

\hypertarget{run-a-different-file}{%
\subsubsection{Run a different file}\label{run-a-different-file}}

Now if we want to process a different expirement, we have to
\textbf{\emph{change our arguments in 6 places!!!}}

Let's use a variable to make it easier to re-run the program on new
data.

\hypertarget{command-version-2}{%
\subsection{Command version 2:}\label{command-version-2}}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{n+nv}{SRRID}\PY{o}{=}SRR5832198
\PY{n+nv}{\PYZdl{}fastp} \PYZhy{}i 01\PYZus{}input/\PY{l+s+si}{\PYZdl{}\PYZob{}}\PY{n+nv}{SRRID}\PY{l+s+si}{\PYZcb{}}\PYZus{}1.fastq       \PYZhy{}I 01\PYZus{}input/\PY{l+s+si}{\PYZdl{}\PYZob{}}\PY{n+nv}{SRRID}\PY{l+s+si}{\PYZcb{}}\PYZus{}2.fastq \PY{l+s+se}{\PYZbs{}}
           \PYZhy{}o 02\PYZus{}output/\PY{l+s+si}{\PYZdl{}\PYZob{}}\PY{n+nv}{SRRID}\PY{l+s+si}{\PYZcb{}}\PYZus{}trim\PYZus{}1.fastq \PYZhy{}O 02\PYZus{}output/\PY{l+s+si}{\PYZdl{}\PYZob{}}\PY{n+nv}{SRRID}\PY{l+s+si}{\PYZcb{}}\PYZus{}trim\PYZus{}2.fastq \PY{l+s+se}{\PYZbs{}}
           \PYZhy{}h 02\PYZus{}output/\PY{l+s+si}{\PYZdl{}\PYZob{}}\PY{n+nv}{SRRID}\PY{l+s+si}{\PYZcb{}}\PYZus{}report.html  \PYZhy{}j 02\PYZus{}output/\PY{l+s+si}{\PYZdl{}\PYZob{}}\PY{n+nv}{SRRID}\PY{l+s+si}{\PYZcb{}}\PYZus{}report.json\PY{l+s+se}{\PYZbs{}}
           \PYZhy{}\PYZhy{}thread \PY{l+m}{1} \PY{l+s+se}{\PYZbs{}}
           \PYZhy{}g \PYZhy{}x \PYZhy{}p
\end{Verbatim}
\end{tcolorbox}

    Modify the value of \texttt{SRRID} to run the command on another
dataset.

    \hypertarget{step-5.-scripting-and-running-a-batch-job}{%
\section{Step 5. Scripting and running a batch
job}\label{step-5.-scripting-and-running-a-batch-job}}

Now we're going to set up the full version of this.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Go back to your file browser and click New-\textgreater Terminal. This
  will open a web-based terminal in a new browser tab.
\item
  Do \texttt{cd\ /scratch/summit/\$USER/DSCI512\_RNAseq\_19\_19\_11}
\item
  Using \emph{nano}, copy the template script below into a new file
  called \texttt{fastp.sbatch}.
\item
  You will submit the job using
  \texttt{sbatch\ -\/-reservation=csuhpc\_nov19\ fastp.sbatch}
\item
  You can check your job status with: \texttt{squeue\ -u\ \$USER} and/or
  \texttt{sacct}
\end{enumerate}

    \hypertarget{a-template-sbatch-script}{%
\subsection{A template SBATCH script}\label{a-template-sbatch-script}}

{[}Not a code cell. Paste into \emph{nano} and modify{]} ***

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#!/usr/bin/bash}
\CommentTok{#SBATCH --nodes=1}
\CommentTok{#SBATCH --ntasks=6}
\CommentTok{#SBATCH --time=0:10:00}
\CommentTok{#SBATCH --qos=normal}
\CommentTok{#SBATCH --partition=shas}
\CommentTok{#SBATCH --output=04_output/fastp.%j.out}


\CommentTok{# 1) Load module}

\CommentTok{# 2) Make shortcut}

\CommentTok{# 3) run the command in a loop for a set of files}
\VariableTok{SRRIDs=}\StringTok{"SRR5832182 SRR5832183"}

\KeywordTok{for} \ExtensionTok{SRRID}\NormalTok{ in }\VariableTok{$SRRIDs}
\KeywordTok{do}
    \CommentTok{# paste Command 2 from above}
\KeywordTok{done}
\end{Highlighting}
\end{Shaded}

\begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

    \hypertarget{my-finished-sbatch-script}{%
\subsection{My finished SBATCH script}\label{my-finished-sbatch-script}}

{[}Not a code cell. Paste into \emph{nano} and modify{]}

\begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#!/usr/bin/bash}
\CommentTok{#SBATCH --nodes=1}
\CommentTok{#SBATCH --ntasks=6}
\CommentTok{#SBATCH --time=0:10:00}
\CommentTok{#SBATCH --qos=normal}
\CommentTok{#SBATCH --partition=shas}
\CommentTok{#SBATCH --output=04_output/fastp.%j.out}

\CommentTok{# 1) Load module:}
\ExtensionTok{module}\NormalTok{ load singularity}

\CommentTok{# 2) Make shortcut}
\VariableTok{fastp=}\StringTok{'singularity exec /projects/dcking@colostate.edu/containers/Summit_RNAseq_container.sif fastp'}

\CommentTok{# 3) run the command in a loop for a set of files}
\VariableTok{SRRIDs=}\StringTok{"SRR5832182 SRR5832183"}

\KeywordTok{for} \ExtensionTok{SRRID}\NormalTok{ in }\VariableTok{$SRRIDs}
\KeywordTok{do}
    \VariableTok{$fastp} \ExtensionTok{-i}\NormalTok{ 01_input/}\VariableTok{$\{SRRID\}}\NormalTok{_1.fastq       -I 01_input/}\VariableTok{$\{SRRID\}}\NormalTok{_2.fastq \textbackslash{}}
\NormalTok{           -o 02_output/}\VariableTok{$\{SRRID\}}\NormalTok{_trim_1.fastq -O 02_output/}\VariableTok{$\{SRRID\}}\NormalTok{_trim_2.fastq \textbackslash{}}
\NormalTok{           -h 02_output/}\VariableTok{$\{SRRID\}}\NormalTok{_report.html  -j 02_output/}\VariableTok{$\{SRRID\}}\NormalTok{_report.json\textbackslash{}}
\NormalTok{           --thread }\VariableTok{$\{SLURM_NTASKS\}}\NormalTok{ \textbackslash{}}
\NormalTok{           -g -x -p  }
\KeywordTok{done}
\end{Highlighting}
\end{Shaded}

\begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

    \hypertarget{my-array-sbatch-script}{%
\subsection{My ARRAY SBATCH script}\label{my-array-sbatch-script}}

{[}Not a code cell. Paste into \emph{nano} and modify{]}

SLURM will submit these jobs in parallel, and so I requested fewer
resources per job. You just have to make sure to match the array
parameter to the way the files are named.

They are SRR5832182 through SRR5832199. So I'll set the array to go from
82 to 99, and just attach it to the rest.

Run like:

\texttt{sbatch\ -\/-reservation=csuhpc\_nov19\ -a\ 82-99\ fastp\_array.sbatch}

\begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#!/usr/bin/bash}
\CommentTok{#SBATCH --nodes=1}
\CommentTok{#SBATCH --ntasks=2}
\CommentTok{#SBATCH --time=0:03:00}
\CommentTok{#SBATCH --qos=normal}
\CommentTok{#SBATCH --partition=shas}

\CommentTok{# run like}
\CommentTok{# sbatch -a 82-99 fastp_array.sbatch}

\CommentTok{# 1) setup:}
\CommentTok{#  a) Load modules}
\CommentTok{#  b) make alias}
\ExtensionTok{module}\NormalTok{ load singularity}

\CommentTok{# 'alias' doesn't work in scripts. Here's an alternative syntax to 'alias':}
\VariableTok{fastp=}\StringTok{'singularity exec /projects/dcking@colostate.edu/containers/Summit_RNAseq_container.sif fastp'}
\CommentTok{# use like:}
\CommentTok{#  $fastp arg1 arg2 ...}


\CommentTok{# 3) Figure out the file root from the job array id}
\CommentTok{# This script must be run like:}
\CommentTok{#  sbatch --array=82-99 fastp_array.sbatch}
\CommentTok{# in order for the IDs to match up to the filenames properly.}

\VariableTok{SRRID=}\StringTok{"SRR58321}\VariableTok{$\{SLURM_ARRAY_TASK_ID\}}\StringTok{"}

\VariableTok{$fastp} \ExtensionTok{-i}\NormalTok{ 01_input/}\VariableTok{$\{SRRID\}}\NormalTok{_1.fastq -I 01_input/}\VariableTok{$\{SRRID\}}\NormalTok{_2.fastq \textbackslash{}}
\NormalTok{    -o 02_output/}\VariableTok{$\{SRRID\}}\NormalTok{_trim_1.fastq   -O 02_output/}\VariableTok{$\{SRRID\}}\NormalTok{_trim_2.fastq \textbackslash{}}
\NormalTok{    -h 02_output/}\VariableTok{$\{SRRID\}}\NormalTok{_report.html    -j 02_output/}\VariableTok{$\{SRRID\}}\NormalTok{_report.json\textbackslash{}}
\NormalTok{    --thread }\VariableTok{$\{SLURM_NTASKS\}}\NormalTok{ \textbackslash{}}
\NormalTok{    -g -x -p  }
\end{Highlighting}
\end{Shaded}

\begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

    \hypertarget{finishing-up}{%
\section{Finishing up}\label{finishing-up}}

\hypertarget{some-notes-about-todays-summit-session}{%
\subsection{Some notes about today's Summit
session}\label{some-notes-about-todays-summit-session}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  The reservation argument only works today during class. Omit it for
  your own work.
\item
  All work on the scratch volume, including my data download, will
  expire after 90 days. If you want to save it, \textbf{copy it to your
  projects directory}, \texttt{/projects/\$USER}, \textbf{not your home
  directory!!!}. You will immediately fill up your quota and can cause
  major issues.
\end{enumerate}

\hypertarget{logging-off}{%
\subsection{Logging off}\label{logging-off}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\item
  Stop the running notebook and terminal
\item
  Shut down the server session (the interactive session).
\item
  Log out
\item
  Click on the \textbf{\emph{stop button}} on the left (light grey
  square inside dark gray circle), hit the X's next to the listed
  things.
\item
  Go to File-\textgreater Hub Control Panel. Click the red button ``Stop
  my server''.
\item
  Click the \textbf{\emph{Logout button}} on the same page.
\end{enumerate}


    % Add a bibliography block to the postdoc
    
    
    
\end{document}
